### PR TITLE
DataManagementApi: add PolicyService and Policy Rest Endpoints(PolicyTransformer and Dto removed)

### LINKED ISSUE #808: DataManagementApi: add `PolicyService` and `PolicyTransformer`
For the implementation of the policy endpoint we need an `PolicyService` (and an `PolicyServiceImpl`) with roughly the following signature:
```java
public interface PolicyService {
  Policy findbyId(String policyId);
  Collection<Policy> query(QuerySpec query);
  Policy delete(String policyId);
  void create(Policy policy);
}
```

In addition, there needs to be an `PolicyTransformer`, that converts an `Policy` into a `PolicyDefinitionDto` and back. This depends on #779 

Important aspects:
- service must provide transactional consistency
- service must perform defensive checks before performing an operation:
  - `delete`: a `Policy` can only be removed if not yet referenced in a `ContractDefinition`
- invoking the transformer is done by the controller, **not** the service.
- transformer does not perform any validation, it simply converts
- depends on #803

### PR DESCRIPTION
**Pull request by diegogomez-zf**
## What this PR changes/adds

This pull requests adds a transformer from Policy to PolicyDefinitionDto (and Vice versa) and a PolicyService 

## Why it does that

 It's necessary for the implementation of the policy endpoint.

## Linked Issue(s)

Closes #808 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    public PolicyServiceImpl(TransactionContext transactionContext, PolicyStore policyStore, ContractDefinitionStore contractDefinitionStore) {
+        this.transactionContext = transactionContext;
+        this.policyStore = policyStore;
+        this.contractDefinitionStore = contractDefinitionStore;
+    }
+
+    @Override
+    public Policy findById(String policyId) {
+        return policyStore.findById(policyId);
```
Comment (ndr-brt) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:42]: please wrap this with the `transactionContext.execute`, same for the `query` method

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            }
+
+            var accessFilter = format("accessPolicy.uid = %s ", policyId);
+            var queryAccessPolicyFilter = QuerySpec.Builder.newInstance().filter(accessFilter).build();
+            var accessDefinitionOnPolicy = contractDefinitionStore.findAll(queryAccessPolicyFilter);
+            if (accessDefinitionOnPolicy.findAny().isPresent()) {
+                return ServiceResult.conflict(format("Policy %s cannot be deleted as it is referenced by at least one accessPolicy", policyId));
+            }
+
+            Policy deleted = policyStore.delete(policyId);
```
Comment (ndr-brt) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:67]:
```suggestion
            var deleted = policyStore.delete(policyId);
```

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            var contractDefinitionOnPolicy = contractDefinitionStore.findAll(queryContractPolicyFilter);
+            if (contractDefinitionOnPolicy.findAny().isPresent()) {
+                return ServiceResult.conflict(format("Policy %s cannot be deleted as it is referenced by at least one contractPolicy", policyId));
+            }
+
+            var accessFilter = format("accessPolicy.uid = %s ", policyId);
+            var queryAccessPolicyFilter = QuerySpec.Builder.newInstance().filter(accessFilter).build();
+            var accessDefinitionOnPolicy = contractDefinitionStore.findAll(queryAccessPolicyFilter);
+            if (accessDefinitionOnPolicy.findAny().isPresent()) {
+                return ServiceResult.conflict(format("Policy %s cannot be deleted as it is referenced by at least one accessPolicy", policyId));
```
Comment (jimmarino) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:64]: nit: "access policy"

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyService.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+public interface PolicyService {
+
+    /**
+     * Returns a policy by its id
+     *
+     * @param policyId id of the policy
+     * @return the policy, null if it's not found
+     */
+    Policy findById(String policyId);
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyService.java:30]: add `@NotNull` and `@Nullable` annotations to methods on interface and implementation?

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyService.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+package org.eclipse.dataspaceconnector.api.datamanagement.policy.service;
+
+import org.eclipse.dataspaceconnector.api.result.ServiceResult;
+import org.eclipse.dataspaceconnector.policy.model.Policy;
+import org.eclipse.dataspaceconnector.spi.query.QuerySpec;
+
+import java.util.Collection;
+
+public interface PolicyService {
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyService.java:22]: consider adding a class javadoc

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformer.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public Class<Policy> getOutputType() {
+        return Policy.class;
+    }
+
+    @Override
+    public @Nullable Policy transform(@Nullable PolicyDefinitionDto object, @NotNull TransformerContext context) {
+        assert object != null;
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformer.java:37]: avoid use of `assert`. Also this contradicts `@Nullable` above.

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformer.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        return Policy.Builder.newInstance().permissions(object.getPermissions())
+                .prohibitions(object.getProhibitions()).duties(object.getObligations())
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformer.java:39]:
```suggestion
                .prohibitions(object.getProhibitions())
                .duties(object.getObligations())
```

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyToPolicyDefinitionDtoTransformer.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public Class<PolicyDefinitionDto> getOutputType() {
+        return PolicyDefinitionDto.class;
+    }
+
+    @Override
+    public @Nullable PolicyDefinitionDto transform(@Nullable Policy object, @NotNull TransformerContext context) {
+        assert object != null;
+        return PolicyDefinitionDto.Builder.newInstance().permissions(object.getPermissions())
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyToPolicyDefinitionDtoTransformer.java:38]: fix wrapping as previous

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyToPolicyDefinitionDtoTransformer.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public Class<PolicyDefinitionDto> getOutputType() {
+        return PolicyDefinitionDto.class;
+    }
+
+    @Override
+    public @Nullable PolicyDefinitionDto transform(@Nullable Policy object, @NotNull TransformerContext context) {
+        assert object != null;
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyToPolicyDefinitionDtoTransformer.java:37]: as previous

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        when(policyStore.findById("policyId")).thenReturn(policy);
+
+        var inserted = policyService.create(policy);
+
+        assertThat(inserted.succeeded()).isFalse();
+    }
+
+    @Test
+    void delete_shouldDeletePolicyIfItsNotReferencedByAnyContractDefinition() {
+        when(contractDefinitionStore.findAll(any())).thenReturn(Stream.empty()).thenReturn(Stream.empty());
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java:88]: nit: second call to `thenReturn` is unnnecessary. Subsequent calls will keep returning the value from the last `thenReturn`

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        assertThat(policy).isNotNull().matches(hasId(assetId));
+    }
+
+    @Test
+    void query_shouldRelyOnPolicyStore() {
+        var policy = createPolicy("policyId");
+        when(policyStore.findAll(any(QuerySpec.class))).thenReturn(Stream.of(policy));
+        var policies = policyService.query(QuerySpec.none());
+
+        assertThat(policies).hasSize(1).first().matches(hasId("policyId"));
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java:62]: as above, simplify to `assertThat(policies).containsExactly(policy);` which is also deeper testing (tests more than just the id field)

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Test
+    void createPolicy_shouldCreatePolicyIfItDoesNotAlreadyExist() {
+        var policy = createPolicy("policyId");
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java:67]: move this to a field? `Policy policy = createPolicy("policyId");`

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformerTest.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+public class PolicyDefinitionDtoToPolicyTransformerTest {
+
+    private final PolicyDefinitionDtoToPolicyTransformer transformer = new PolicyDefinitionDtoToPolicyTransformer();
+
+    @Test
+    void inputOutputType() {
+        assertThat(transformer.getInputType()).isInstanceOf(PolicyDefinitionDto.class);
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformerTest.java:37]: `isEqualTo`. A class is not an instance of itself

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformerTest.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        ArrayList<Duty> obligationsTest = new ArrayList<>();
+
+        var context = mock(TransformerContext.class);
+        var policyDefinitionDto = PolicyDefinitionDto.Builder.newInstance()
+                .inheritsFrom("inheritant")
+                .assigner("the tester")
+                .assignee("the tested")
+                .target("the target")
+                .extensibleProperties(Map.of("key", "value"))
+                .permissions(permissionsTest)
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformerTest.java:54]:
```suggestion
                .permissions(List.of())
```
also others

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformerTest.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                .prohibitions(prohibitionsTest)
+                .obligations(obligationsTest)
+                .build();
+
+        var policy = transformer.transform(policyDefinitionDto, context);
+
+        SoftAssertions softAssertions = new SoftAssertions();
```
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/transform/PolicyDefinitionDtoToPolicyTransformerTest.java:61]: See TransferProcessToTransferProcessDtoTransformerTest for a simpler and more robust way.

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java (hunk 2 above)
Comment (algattik) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:67]: needs upstream merge
```suggestion
            Policy deleted = policyStore.deleteById(policyId);
```

**Review by algattik — CHANGES_REQUESTED** [review]
some suggestions

File: extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java (hunk 10 above)
Comment (ndr-brt) [reply extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java:88]: @algattik the problem is that the Stream can be consumed once, and the `thenReturn` returns the same instance, so removing the second `thenReturn` will throw exception

Comment (diegogomez-zf) [reply extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java:88]: @algattik without the second thenReturn I received the exception "java.lang.IllegalStateException: stream has already been operated upon or closed" 

I can do however .thenReturn(Stream.empty(), Stream.empty()) which also works and only uses one thenReturn

Comment (algattik) [reply extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java:88]: Interesting :)

**Review by ndr-brt — APPROVED** [review]

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public Collection<Policy> query(QuerySpec query) {
+        return transactionContext.execute(() ->
+                policyStore.findAll(query).collect(toList()));
+    }
+
+    @Override
+    public ServiceResult<Policy> delete(String policyId) {
+        return transactionContext.execute(() -> {
+            var contractFilter = format("contractPolicy.uid = %s ", policyId);
```
Comment (paullatzelsperger) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:55]: stuff that is not transactional, such as formatting strings, creating objects, etc. should be done outside transactions

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
After some discussions with @jimmarino we concluded that there is in fact no need for a `PolicyDefinitionDto`, since the `Policy` is already intended to be serializable. In addition, we'd have to also create DTOs for `Permission`, `Prohibition` and `Duty`, which would be painful.

I know I specified it differently, but could you remove the `PolicyDefinitionDto` and the respective transformers again? I'm sorry, we should have discussed this before your PR. 

BTW the transformer tests are failing due to incorrect assertions, but that will get solved implicitly, once the dto and transformers are gone.

File: extensions/api/data-management/policydefinition/build.gradle.kts [hunk 17]
```diff
@@ -25,6 +25,9 @@ dependencies {
     implementation(project(":common:util"))
     implementation(project(":extensions:api:api-core"))
     implementation(project(":extensions:api:data-management:api-configuration"))
+    implementation(project(":spi:policy-spi"))
+    implementation(project(":spi:contract-spi"))
+    implementation(project(":extensions:transaction:transaction-spi"))
 
     implementation("jakarta.ws.rs:jakarta.ws.rs-api:${rsApi}")
     testImplementation(testFixtures(project(":launchers:junit")))
```
Comment (paullatzelsperger) [inline extensions/api/data-management/policydefinition/build.gradle.kts:33]: the following deps are not needed. pls remove:
```kotlin
    testImplementation(testFixtures(project(":launchers:junit")))
    testImplementation(project(":extensions:in-memory:contractdefinition-store-memory"))
```

Comment (diegogomez-zf) [reply extensions/api/data-management/policydefinition/build.gradle.kts]: Hi Paul, without the junit I cannot import org.eclipse.dataspaceconnector.junit.launcher.EdcExtension, that I use in my IntegrationTest.

Without the contractDefinition-store-memory I cannot check if a Policy is already part of a contract definition in order to be able to delete it.

**Review by ndr-brt — APPROVED** [review]
LGTM

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+    public PolicyServiceImpl(TransactionContext transactionContext, PolicyStore policyStore, ContractDefinitionStore contractDefinitionStore) {
+        this.transactionContext = transactionContext;
+        this.policyStore = policyStore;
+        this.contractDefinitionStore = contractDefinitionStore;
+    }
+
+    @Override
+    public Policy findById(String policyId) {
+        return transactionContext.execute(() ->
```
Comment (jimmarino) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:43]: Please use braces

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+                policyStore.findById(policyId));
+    }
+
+    @Override
+    public @NotNull Collection<Policy> query(QuerySpec query) {
+        return transactionContext.execute(() ->
```
Comment (jimmarino) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImpl.java:49]: Please use braces

File: extensions/api/data-management/policydefinition/build.gradle.kts (hunk 17 above)
Comment (paullatzelsperger) [reply extensions/api/data-management/policydefinition/build.gradle.kts]: Mmh, i see. Leave it in then. Strange thing is, that i deleted those lines, re-ran the test an it worked? That's why I added this comment in the first place. Maybe it was a caching issue. 
Alright, disregard the comment then.

**Pull request merged**